deals/store.go

```diff
@@ -0,0 +1,195 @@
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	err := s.ds.Delete(makePendingDealKey(dr.DealInfo.ProposalCid))
+	if err != nil {
+		return fmt.Errorf("deleting PendingDeal: %s", err)
+	}
+	buf, err := json.Marshal(dr)
+	if err != nil {
+		return fmt.Errorf("marshaling DealRecord: %s", err)
+	}
+	if err := s.ds.Put(makeFinalDealKey(dr.DealInfo.ProposalCid), buf); err != nil {
+		return fmt.Errorf("put DealRecord: %s", err)
+	}
+	return nil
+}
+
+func (s *store) getFinalDeals() ([]DealRecord, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	q := query.Query{Prefix: dsBaseStorageFinal.String()}
+	res, err := s.ds.Query(q)
+	if err != nil {
+		return nil, fmt.Errorf("executing query: %s", err)
+	}
+	defer func() {
+		if err := res.Close(); err != nil {
+			log.Errorf("closing query result: %s", err)
+		}
+	}()
+
+	var ret []DealRecord
+	for r := range res.Next() {
+		var dealRecord DealRecord
+		if err := json.Unmarshal(r.Value, &dealRecord); err != nil {
+			return nil, fmt.Errorf("unmarshaling query result: %s", err)
+		}
+		ret = append(ret, dealRecord)
+	}
+	sort.Slice(ret, func(i, j int) bool {
+		if ret[i].DealInfo.ActivationEpoch < ret[j].DealInfo.ActivationEpoch {
+			return true
+		}
+		if ret[i].DealInfo.ActivationEpoch > ret[j].DealInfo.ActivationEpoch {
+			return false
+		}
+		return ret[i].Time < ret[j].Time
+	})
+	return ret, nil
+}
+
+func (s *store) putRetrieval(rr RetrievalRecord) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	buf, err := json.Marshal(rr)
+	if err != nil {
+		return fmt.Errorf("marshaling RetrievalRecord: %s", err)
+	}
+	if err := s.ds.Put(makeRetrievalKey(rr), buf); err != nil {
+		return fmt.Errorf("put RetrievalRecord: %s", err)
+	}
+	return nil
+}
+
+func (s *store) getRetrievals() ([]RetrievalRecord, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	q := query.Query{Prefix: dsBaseRetrieval.String()}
+	res, err := s.ds.Query(q)
+	if err != nil {
+		return nil, fmt.Errorf("executing query: %s", err)
+	}
+	defer func() {
+		if err := res.Close(); err != nil {
+			log.Errorf("closing query result: %s", err)
+		}
+	}()
+
+	var ret []RetrievalRecord
+	for r := range res.Next() {
+		var rr RetrievalRecord
+		if err := json.Unmarshal(r.Value, &rr); err != nil {
+			return nil, fmt.Errorf("unmarshaling query result: %s", err)
+		}
+		ret = append(ret, rr)
+	}
+	sort.Slice(ret, func(i, j int) bool {
+		return ret[i].Time < ret[j].Time
+	})
+	return ret, nil
+}
+
+func (s *store) close() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	err := s.ds.Close()
+	return err
+}
+
+func makePendingDealKey(c cid.Cid) datastore.Key {
+	return dsBaseStoragePending.ChildString(c.String())
+}
+
+func makeFinalDealKey(c cid.Cid) datastore.Key {
+	return dsBaseStorageFinal.ChildString(c.String())
+}
+
+func makeRetrievalKey(rr RetrievalRecord) datastore.Key {
+	str := fmt.Sprintf("%v%v%v%v", rr.Time, rr.Addr, rr.RetrievalInfo.Miner, rr.RetrievalInfo.PieceCID.String())
+	sum := md5.Sum([]byte(str))
```

Comment:
needed a unique key so just built one like this, no idea if there is a better way

Reply:
I guess should be fine. But maybe all this can be avoided by generating a `uuid` and using that as the key?
Since this key isn't that useful anyway to do a Query or similar.

Reply:
Actually, I think my next PR will rethink the keys used to make it easier to query by address etc. So I'll leave this for now knowing that it is going to change anyway.

Reply:
Cool 👍